man/recins.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.40.10.
.TH RECINS "1" "April 2014" "recins 1.7" "User Commands"
.SH NAME
recins \- insert records in a recfile
.SH SYNOPSIS
.B recins
[\fIOPTION\fR]... [\fI-t TYPE\fR] [\fI-n NUM | -e RECORD_EXPR | -q STR | -m NUM\fR] [\fI(-f NAME -v STR) | -r RECDATA\fR]... [\fIFILE\fR]
.SH DESCRIPTION
Insert new records in a rec database.
.TP
\fB\-f\fR, \fB\-\-field\fR=\fISTR\fR
field name; should be followed by a \fB\-v\fR.
.TP
\fB\-v\fR, \fB\-\-value\fR=\fISTR\fR
field value; should be preceded by an \fB\-f\fR.
.TP
\fB\-r\fR, \fB\-\-record\fR=\fISTR\fR
record that will be inserted in the file.
.TP
\fB\-\-force\fR
insert the record even if it is violating
record restrictions.
.TP
\fB\-\-no\-external\fR
don't use external descriptors.
.TP
\fB\-\-no\-auto\fR
don't insert auto generated fields.
.TP
\fB\-\-verbose\fR
give a detailed report if the integrity check
fails.
.TP
\fB\-s\fR, \fB\-\-password\fR=\fISTR\fR
encrypt confidential fields with the given password.
.TP
\fB\-\-help\fR
print a help message and exit.
.TP
\fB\-\-version\fR
show version and exit.
.SS "Record selection options:"
.TP
\fB\-i\fR, \fB\-\-case\-insensitive\fR
make strings case\-insensitive in selection
expressions.
.TP
\fB\-t\fR, \fB\-\-type\fR=\fITYPE\fR
operate on records of the specified type only.
.TP
\fB\-e\fR, \fB\-\-expression\fR=\fIRECORD_EXPR\fR
selection expression.
.TP
\fB\-q\fR, \fB\-\-quick\fR=\fISTR\fR
select records with fields containing a string.
.TP
\fB\-n\fR, \fB\-\-number\fR=\fINUM\fR,...
select specific records by position, with ranges.
.TP
\fB\-m\fR, \fB\-\-random\fR=\fINUM\fR
select a given number of random records.
.PP
If no FILE is specified then the command acts like a filter, getting
the data from standard input and writing the result to standard output.
.SH AUTHOR
Written by Jose E. Marchesi.
.SH "REPORTING BUGS"
Report bugs to: bug\-recutils@gnu.org
.br
GNU recutils home page: <http://www.gnu.org/software/recutils/>
.br
General help using GNU software: <http://www.gnu.org/gethelp/>
.SH COPYRIGHT
Copyright \(co 2010, 2011, 2012, 2013, 2014 Jose E. Marchesi.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
.SH "SEE ALSO"
The full documentation for
.B recins
is maintained as a Texinfo manual.  If the
.B info
and
.B recins
programs are properly installed at your site, the command
.IP
.B info recutils
.PP
should give you access to the complete manual.


NEWS
This NEWS file records noteworthy changes, very tersely.
See the manual for detailed information.

  Copyright 2010-2014 Jose E. Marchesi.

  Copying and distribution of this file, with or without modification,
  are permitted in any medium without royalty provided the copyright
  notice and this notice are preserved.

 Version 1.7 (25 March 2014)
* Utilities:
  . rec2csv now supports a -d option to change the delimiter
    character in CSV files.
  . It is now possible to specify several recfiles in the
    command line containing anonymous records.
* Format:
  . Support for restricting the fields that can appear in records
    with the new %allowed special field.
  . If a tab separates a field name and a field value is now
    ignored and not considered part of the field value.
* librec:
  . The current version of the library is now 1.
* Emacs mode:
  . It now possible to jump several records while navigating
    using a prefix argument to the next/previous commands.
* Documentation:
  . The examples in the manual have been improved, and well as many
    other aspects: formatting, English, etc.  Special thanks to Karl
    Berry for pointing out the problems and suggesting solutions.
* Internal cleanup and code factorization.
* Many, many, many bug fixes :D

 Version 1.6 (03 November 2013)
* Utilities:
  . Support for joins and foreign keys.
  . Grouping records by one or more fields.
  . Sorting recors by several fields.
  . Specifying rewrite rules in field expressions.
  . Writing field values in rows.
  . Dot-notation in selection expressions.
  . Support for the >= and <= operators in selection expressions.
  . Aggregated functions: sum, min, max, avg.
  . New bash loadable builtin `readrec'.
* Format:
  . New UUID field type.
  . User provided arbitrary constraints with %constraint.
  . Pre-defined constants MIN and MAX can now be used in ranges.
* librec:
  . New high-level API: rec_db_query, rec_db_insert, rec_db_delete and
    rec_db_set.
  . Documentation in the rec.h header file extended and improved.
* Emacs mode:
  . Texinfo manual describing the usage of rec-mode.
  . Support for a mode hook.
  . Comments are now supported in records.
  . add-change-log-entry in a recfile now does the right thing.
  . Commands for copying/killing fields and records.
  . Support for undo in navigation mode.
  . Field folding.
  . Editing methods for the several field types.
  . Statistics.
* Documentation:
  . The user manual has been restructured and expanded.
* Internal cleanup and code factorization.
* Many, many, many bug fixes :D

 Version 1.5 (13 January 2012)
* The utilities will now ask interactively for a password if it was
  not provided with the -s command line option.  This avoids security
  problems related to shell history files.
* Support for octal and hexadecimal numbers has been added.  They can
  be used in both the records and selection expressions.
* It is now possible to select a given number of random records in
  many of the utilities using the -m command line option.
* The -n option now accepts a list of indexes, supporting ranges.
* The new -U (uniq) option for recsel removes duplicated fields in the
  output records.
* The new -q (quick) option allows to quickly search for the desired
  record without having to provide a complete selection expression.
* Auto generated fields are now considered integers by default.  This
  avoids repetitive patterns in record descriptors involving %auto and
  %type.
* Tab characters are now allowed in blank lines betwwen records.
* The API in rec.h is now better documented with comments, and
  improved.
* recfix now exits with an error status if there is a parse error in
  some input file.
* The usage of the internal data structures has been _vastly_
  improved, resulting in a much faster operation.
* Internal cleanup and code factorization.
* Many, many, many bug fixes :D

 Version 1.4 (03 November 2011)
* Support for encryption with the new %confidential special field.
* Support for named types with the new %typedef special field.
* Support for sorting with the new %sort special field.
* Recutils support for org-babel is now included in the distribution
  as ob-rec.el.  It allows to read rec data into org documents in both
  raw and table formats.
* It is now possible to use double quotes for delimiting string
  literals in selection expressions, as well as single quotes.
  Unescapted newline characters are also supported.
* The manual has been vastly improved.
* recfix now supports several operations: check, sort, encrypt and
  decrypt.
* recins now generates parseable locale independent dates when
  generating time-stamps in auto-fields.
* Lots of bug fixes.

 Version 1.3 (11 February 2011)
* A quite bad (and embarrassing) bug introduced in 1.2 has been
  fixed.  The bug was causing recins, recdel and recset to report
  random integrity failures, due to a uninitialized variable.
* New special field %size, that allows to define constraints in the
  size of record sets.
* Several fixes in the user manual.
* Little portability issue that prevented one of the tests to compile
  in macos x has been fixed.
* Translations updates.

 Version 1.2 (4 February 2011)
* New utility rec2csv, to export rec data to csv files.
* Support for auto-incremented fields and timestamps with the new
  %auto field in record descriptors.
* Support for comments in enum type descriptions.
* recset now supports a new "set or add field" operation, implementing
  the common case of setting the value of a field in case it exists,
  or else add it to the record.
* The flavor of regular expressions accepted in both selection
  expressions and type descriptions is now documented in the user
  manual.
* Translations for German, Finnish, Swedish and Dutch.
* recdel now performs an integrity check before to complete the
  deletion.
* The utilities no longer "expand" the contents of remote descriptors
  in the local ones.

utils/recins.c
/* -*- mode: C -*-
 *
 *       File:         recins.c
 *       Date:         Mon Dec 28 08:54:38 2009
 *
 *       GNU recutils - recins
 *
 */

/* Copyright (C) 2009-2013 Jose E. Marchesi */

/* This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <config.h>

#include <getopt.h>
#include <string.h>
#include <stdlib.h>
#include <xalloc.h>
#include <gettext.h>
#include <errno.h>
#include <time.h>
#include <locale.h>
#define _(str) gettext (str)
#include <base64.h>

#include <rec.h>
#include <recutl.h>

/* Forward declarations.  */
bool recins_insert_record (rec_db_t db, char *type, rec_record_t record);
void recins_parse_args (int argc, char **argv);

/*
 * Global variables
 */

char         *recutl_type      = NULL;
rec_sex_t     recutl_sex       = NULL;
char         *recutl_sex_str   = NULL;
char         *recutl_quick_str = NULL;
bool          recutl_insensitive = false;
rec_record_t  recins_record    = NULL;
char         *recins_file      = NULL;
bool          recins_force     = false;
bool          recins_verbose   = false;
bool          recins_external  = true;
bool          recins_auto      = true;
char         *recins_password  = NULL;
size_t        recutl_random    = 0;

/*
 * Command line options management
 */

enum
{
  COMMON_ARGS,
  RECORD_SELECTION_ARGS,
  NAME_ARG,
  VALUE_ARG,
  FORCE_ARG,
  VERBOSE_ARG,
  NO_EXTERNAL_ARG,
  RECORD_ARG,
#if defined REC_CRYPT_SUPPORT
  PASSWORD_ARG,
#endif
  NO_AUTO_ARG
};

static const struct option GNU_longOptions[] =
  {
    COMMON_LONG_ARGS,
    RECORD_SELECTION_LONG_ARGS,
    {"type", required_argument, NULL, TYPE_ARG},
    {"name", required_argument, NULL, NAME_ARG},
    {"value", required_argument, NULL, VALUE_ARG},
    {"force", no_argument, NULL, FORCE_ARG},
    {"verbose", no_argument, NULL, VERBOSE_ARG},
    {"no-external", no_argument, NULL, NO_EXTERNAL_ARG},
    {"record", required_argument, NULL, RECORD_ARG},
    {"no-auto", no_argument, NULL, NO_AUTO_ARG},
#if defined REC_CRYPT_SUPPORT
    {"password", required_argument, NULL, PASSWORD_ARG},
#endif
    {NULL, 0, NULL, 0}
  };

/*
 * Functions.
 */

void
recutl_print_help (void)
{
  /* TRANSLATORS: --help output, recins synopsis.
     no-wrap */
  printf (_("\
Usage: recins [OPTION]... [-t TYPE] [-n NUM | -e RECORD_EXPR | -q STR | -m NUM] [(-f NAME -v STR) | -r RECDATA]... [FILE]\n"));

  /* TRANSLATORS: --help output, recins short description.
     no-wrap */
  fputs (_("\
Insert new records in a rec database.\n"), stdout);

  puts ("");
  /* TRANSLATORS: --help output, recins arguments.
     no-wrap */
  fputs (_("\
  -f, --field=STR                     field name; should be followed by a -v.\n\
  -v, --value=STR                     field value; should be preceded by an -f.\n\
  -r, --record=STR                    record that will be inserted in the file.\n\
      --force                         insert the record even if it is violating\n\
                                        record restrictions.\n\
      --no-external                   don't use external descriptors.\n\
      --no-auto                       don't insert auto generated fields.\n\
      --verbose                       give a detailed report if the integrity check\n\
                                        fails.\n"), stdout);

#if defined REC_CRYPT_SUPPORT
  /* TRANSLATORS: --help output, encryption related options.
     no-wrap */
  fputs (_("\
  -s, --password=STR                  encrypt confidential fields with the given password.\n"),
         stdout);
#endif

  recutl_print_help_common ();

  puts ("");
  recutl_print_help_record_selection ();

  puts ("");
  /* TRANSLATORS: --help output, notes on recins.
     no-wrap */
  fputs (_("\
If no FILE is specified then the command acts like a filter, getting\n\
the data from standard input and writing the result to standard output.\n"), stdout);

  puts ("");
  recutl_print_help_footer ();
}


void recins_parse_args (int argc,
                        char **argv)
{
  int ret;
  char c;
  rec_field_t field = NULL;
  char *field_name = NULL;
  rec_record_t provided_record;

  while ((ret = getopt_long (argc,
                             argv,
                             RECORD_SELECTION_SHORT_ARGS
                             ENCRYPTION_SHORT_ARGS
                             "f:v:r:",
                             GNU_longOptions,
                             NULL)) != -1)
    {
      c = ret;
      switch (c)
        {
          COMMON_ARGS_CASES
          RECORD_SELECTION_ARGS_CASES
        case FORCE_ARG:
          {
            recins_force = true;
            break;
          }
        case VERBOSE_ARG:
          {
            recins_verbose = true;
            break;
          }
        case NAME_ARG:
        case 'f':
          {
            if (field != NULL)
              {
                recutl_fatal (_("a -f should be followed by a -v\n"));
                exit (EXIT_FAILURE);
              }

            if (recins_record == NULL)
              {
                recins_record = rec_record_new ();
                rec_record_set_source (recins_record, "cmdli");
                rec_record_set_location (recins_record, 0);
              }

            if (!rec_field_name_p (optarg))
              {
                recutl_fatal (_("invalid field name %s.\n"), optarg);
              }
            
            field = rec_field_new (optarg, "foo");
            break;
          }
        case VALUE_ARG:
        case 'v':
          {
            if (field == NULL)
              {
                recutl_fatal (_("a -v should be preceded by a -f\n"));
              }

            rec_field_set_value (field, optarg);
            rec_mset_append (rec_record_mset (recins_record), MSET_FIELD, (void *) field, MSET_ANY);

            field = NULL;
            break;
          }
        case NO_EXTERNAL_ARG:
          {
            recins_external = false;
            break;
          }
        case NO_AUTO_ARG:
          {
            recins_auto = false;
            break;
          }
#if defined REC_CRYPT_SUPPORT
        case PASSWORD_ARG:
        case 's':
          {
            if (recins_password != NULL)
              {
                recutl_fatal (_("more than one password was specified\n"));
              }

            recins_password = xstrdup (optarg);
            break;
          }
#endif
        case RECORD_ARG:
        case 'r':
          {
            /* Parse the provided record and put it in recins_record.  */
            provided_record = rec_parse_record_str (optarg);
            if (!provided_record)
              {
                recutl_fatal (_("error while parsing the record provided by -r\n"));
              }

            if (recins_record)
              {
                /* Append the fields in provided_record into
                   recins_record.  */
                
                rec_mset_iterator_t iter = rec_mset_iterator (rec_record_mset (provided_record));
                while (rec_mset_iterator_next (&iter, MSET_FIELD, (const void **) &field, NULL))
                  {
                    rec_mset_append (rec_record_mset (recins_record), MSET_FIELD, (void *) rec_field_dup (field), MSET_ANY);
                    field = NULL;
                  }
                rec_mset_iterator_free (&iter);

                rec_record_destroy (provided_record);
                provided_record = NULL;
              }
            else
              {
                recins_record = provided_record;
              }

            break;
          }
        default:
          {
            exit (EXIT_FAILURE);
          }
        }
    }

  if (field != NULL)
    {
      recutl_fatal (_("please provide a value for the field %s\n"), field_name);
    }

  /* Read the name of the file where to make the insertions.  */
  if (optind < argc)
    {

      if ((argc - optind) != 1)
        {
          recutl_print_help ();
          exit (EXIT_FAILURE);
        }

      recins_file = argv[optind++];
    }
}

void
recins_add_new_record (rec_db_t db)
{
  int flags = 0;

#if defined REC_CRYPT_SUPPORT

  /* Get the password interactively from the user if some field is
     declared as confidential in the requested record set.  */

  {
    rec_rset_t rset;
    rec_fex_t confidential_fields;

    if (recutl_type)
      {
        rset = rec_db_get_rset_by_type (db, recutl_type);
    
        if (rset)
          {
            confidential_fields = rec_rset_confidential (rset);
            if (!confidential_fields)
              recutl_out_of_memory ();
            
            if (rec_fex_size (confidential_fields) > 0)
              {
                if (!recins_password && recutl_interactive ())
                  {
                    recins_password = recutl_getpass (true);
                    if (!recins_password)
                      {
                        recutl_fatal ("not in an interactive terminal.\n");
                      }
                  }
                
                /* Passwords can't be empty.  */
                
                if (recins_password && (strlen (recins_password) == 0))
                  {
                    free (recins_password);
                    recins_password = NULL;
                  }
              }
          }
      }
  }

#endif /* REC_CRYPT_SUPPORT */

  /* Set flags flags and call the library to perform the
     requested insertion/replacement operation.  */

  if (recutl_insensitive)
    {
      flags = flags | REC_F_ICASE;
    }

  if (!recins_auto)
    {
      flags = flags | REC_F_NOAUTO;
    }

  if (!rec_db_insert (db,
                      recutl_type,
                      recutl_index (),
                      recutl_sex,
                      recutl_quick_str,
                      recutl_random,
                      recins_password,
                      recins_record,
                      flags))
    recutl_out_of_memory ();

  /* Check for the integrity of the resulting database.  */

  if (!recins_force && db)
    {
      recutl_check_integrity (db, recins_verbose, recins_external);
    }
}

int
main (int argc, char *argv[])
{
  rec_db_t db;

  recutl_init ("recins");

  recins_parse_args (argc, argv);

  db = recutl_read_db_from_file (recins_file);
  if (!db)
    {
      /* Create an empty database.  */
      db = rec_db_new ();
    }
  recins_add_new_record (db);

  if (!recutl_file_is_writable (recins_file))
    {
      recutl_error (_("file %s is not writable.\n"), recins_file);
      return EXIT_FAILURE;
    }
  recutl_write_db_to_file (db, recins_file);

  return EXIT_SUCCESS;
}

/* End of recins.c */
